wagtail/contrib/settings/context_processors.py

```diff
@@ -40,33 +38,46 @@ def __missing__(self, model_name):
         self[model_name] = value = self.get_setting(model_name)
         return value
 
+    def __str__(self):
+        return "SettingModuleProxy({0})".format(self.app_label)
+
     def get_setting(self, model_name):
         """
         Get a setting instance
         """
         Model = registry.get_by_natural_key(self.app_label, model_name)
         if Model is None:
-            return None
+            raise RuntimeError(
+                f"Could not find model matching `{self.app_label}.{model_name}`."
+            )
 
-        if isinstance(self.request_or_site, Site):
-            return Model.for_site(self.request_or_site)
-        # Utilises cached value on request if set
-        return Model.for_request(self.request_or_site)
+        if issubclass(Model, BaseGenericSetting):
+            return Model.get_or_create()
```

Comment:
Maybe we could provide `self.request_or_site` as a 'cache target' here, and `get_or_create` could look for an 'already fetched' instance on there?

Reply:
@kbayliss any thoughts on this? Would seem like an easy addition, and the caching behaviour would bring things more in line with site settings when referencing the same settings in different templates.